### ISSUE TITLE
Allow Duck-typing checks for conditional API compat logic

### ISSUE THREAD

**Issue by amilajack** [amilajack/eslint-plugin-compat#26]
**UPDATE: Here's a [related discussion in the eslint-typescript repo](https://github.com/typescript-eslint/typescript-eslint/issues/1380) that's worth looking into**

Allow checking for if lib or property is defined. Disallow calling the method
```js
// Checking if objects are defined
if (typeof ServiceWorker === 'undefined') {} // ✅ allowed

// Checking if properties/methods are defined
if (String.prototype.includes) {} // ✅ allowed

// Checking if properties/methods are defined and
// subsequently invoking/accessing them
if (String.prototype.includes) { // ✅ allowed
  if ('some'.includes('e')) { // ✅ allowed
    // do some action
  }
}

// ✅ allowed
if ("serviceWorker" in navigator) {
  navigator.serviceWorker...
}

// ✅ allowed
if (String.prototype.includes && 'some'.includes('e')) {}
if ('foo'.includes && 'some'.includes('e')) {}

// ❌ disallowed
if ('some'.includes('e')) {}

// ❌ disallowed
if (!'some'.includes('e')) {
  'some'.includes('e')
}
```

**Comment by ljharb** [comment]
Wouldn't `ServiceWorker` throw a ReferenceError when it's not defined, unless it's used with `typeof`?

**Comment by ljharb** [comment]
In other words, duck-typing should only be allowed when chained off of a definitely-present object.

**Comment by amilajack** [comment]
Oh yea, that's right. I've updated the rfc the initial comment to reflect this. Thanks for letting me know about this!

**Comment by deecewan** [comment]
what would happen in this case:

```js
if (String.prototype.includes) {
  if ('some'.includes('e')) {
    // do some action
  }
}
```

In this case, you've validated that the function exists before using it.

**Comment by amilajack** [comment]
@deecewan Sorry for late response. Somehow I missed the notification for this. That should be allowed. I just added this example to my initial comment

**Comment by TrevorBurnham** [comment]
@amilajack Say, we're very interested in this feature at my company. Have you made any headway? If you need help, me and some of my colleagues would be happy to roll up our sleeves!

**Comment by SalimBensiali** [comment]
Hi @amilajack, I am also interested in this feature, has any progress been made since? Would you like any contribution? Cheers

**Comment by amilajack** [comment]
@SalimBensiali I always love contributions. Unfortunately I don't have enough time to contribute to this project. Would really love if anyone could work on this!

**Comment by SalimBensiali** [comment]
No worries @amilajack , is there an initial branch, POC? Is this issue here the one source of truth? Happy to look at this. Cheers.

**Comment by amilajack** [comment]
Nope, nothing eyt
